File: Refinement/Framework/RefinementEngine.ts
import { analysisTypes, AnalyzedData, IGameDataAnalyzer, ITrialCombination } from "../../Contracts/AnalysisContracts.js"
import { ScrapedData } from "../../Contracts/ExtractionContracts.js"

export class GameDataAnalyzer<AnalyzedObject extends object> {
    constructor(
        public analyzer : IGameDataAnalyzer<AnalyzedObject>,
        public trialCombination : ITrialCombination<AnalyzedObject>,
    ) 
    {}
}

export const AnalyzeData = (uniqueTrials : ScrapedData[], analysisFunctions : GameDataAnalyzer<analysisTypes>[], analyzedData : AnalyzedData) => {
    analysisFunctions.forEach((analysisFunction)=>{
        const trialData = uniqueTrials.map(trial=>analysisFunction.analyzer(trial))

        // console.log(trialData)

        analysisFunction.trialCombination(trialData, analyzedData)
    })

    return analyzedData
}

File: 1_Analysis/Framework/RefinementEngine.py
from typing import Callable, Generic, List, Any, TypeVar

from Contracts.ExtractionContracts import ScrapedData
from Contracts.AnalysisContracts import AnalyzedData

T = TypeVar('T')


class GameDataAnalyzer(Generic[T]):
    analyzer: Callable[[ScrapedData], T]
    trial_combination: Callable[[List[T], AnalyzedData], None]

    def __init__(
        self,
        analyzer: Callable[[ScrapedData], T],
        trial_combination: Callable[[List[T], AnalyzedData], None],
    ) -> None:
        self.analyzer = analyzer
        self.trial_combination = trial_combination


def analyze_data(
    unique_trials: List[ScrapedData],
    analysis_functions: List[GameDataAnalyzer[Any]],
    analyzed_data: AnalyzedData,
) -> AnalyzedData:
    for analysis_function in analysis_functions:
        trial_data: List[Any] = [analysis_function.analyzer(trial) for trial in unique_trials]
        analysis_function.trial_combination(trial_data, analyzed_data)

    return analyzed_data
